asmcomp/arm64/emit.mlp

```diff
@@ -1021,16 +1068,57 @@ let fundecl fundecl =
   let num_call_gc, num_check_bound =
     num_call_gc_and_check_bound_points env
   in
+
+  (* Dynamic stack checking *)
+  let stack_threshold_size = Config.stack_threshold * 8 in (* bytes *)
+  let { max_stack_size; contains_nontail_calls } =
+    preproc_fun env fundecl.fun_body fundecl.fun_name
+  in
+  let stack_check_size = ref 0 in
+  let handle_overflow = ref None in
+  if contains_nontail_calls || max_stack_size >= stack_threshold_size then begin
+    let overflow = new_label () and ret = new_label () in
+    let threshold_offset = Domainstate.stack_ctx_words * 8 + stack_threshold_size in
+    let s = max_stack_size + threshold_offset in
+    let offset = Domainstate.(idx_of_field Domain_current_stack) * 8 in
+    `  ldr {emit_reg reg_tmp1}, [{emit_reg reg_domain_state_ptr}, #{emit_int offset}]\n`;
+    `  add {emit_reg reg_tmp1}, {emit_reg reg_tmp1}, #{emit_int s}\n`;
+    `  cmp sp, {emit_reg reg_tmp1}\n`;
+    `  bcc {emit_label overflow}\n`;
+    `{emit_label ret}:\n`;
+    handle_overflow := Some (overflow, ret);
+    stack_check_size := 5
+  end;
+
   let max_out_of_line_code_offset =
+    !stack_check_size +
     max_out_of_line_code_offset ~num_call_gc
       ~num_check_bound
   in
+
   BR.relax fundecl ~max_out_of_line_code_offset;
+
   emit_all env fundecl.fun_body;
   List.iter emit_call_gc env.call_gc_sites;
   List.iter emit_call_bound_error env.bound_error_sites;
   assert (List.length env.call_gc_sites = num_call_gc);
   assert (List.length env.bound_error_sites = num_check_bound);
+
+  begin match !handle_overflow with
+  | None -> ()
+  | Some (overflow,ret) -> begin
+      `{emit_label overflow}:\n`;
+      (* Pass the desired stack size on the stack, since all of the
+        argument-passing registers may be in use. *)
+      let s = (Config.stack_threshold + max_stack_size / 8) in
```

Comment:
If `Config.stack_threshold` is removed, then `caml_try_realloc_stack` will have to include that in the `required_space` computation.